src/org/sosy_lab/common/collect/OrderStatisticSet.java

```diff
@@ -0,0 +1,113 @@
+/*
+ *  SoSy-Lab Common is a library of useful utilities.
+ *  This file is part of SoSy-Lab Common.
+ *
+ *  Copyright (C) 2007-2015  Dirk Beyer
+ *  All rights reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.sosy_lab.common.collect;
+
+import java.util.SortedSet;
+
+/**
+ * A {@link SortedSet} that allows two additional operations: receiving (and deleting) an element by
+ * its <i>rank</i>, and getting the rank of an element.
+ *
+ * <p>Implementations should adhere to all contracts of the <code>SortedSet</code> interface.
+ *
+ * <p>Implementing classes should provide two means for comparing elements:
+ *
+ * <ol>
+ *   <li>Using the natural ordering of the elements. In this case, all elements of the set have to
+ *       implement the {@link Comparable} interface.
+ *   <li>Using a {@link java.util.Comparator Comparator} to create an order over the elements of the
+ *       set.
+ * </ol>
+ *
+ * In both cases, the used compare-method should be consistent with <code>equals</code>, i.e.,
+ * <code>compare(a, b) == 0  =&gt;  a.equals(b)</code>, so that the contract provided by {@link
+ * java.util.Set Set} is fulfilled. Implementations of this interface should still work if the used
+ * compare-method is not consistent with <code>equals</code>.
+ *
+ * <p>If the used compare-method is not consistent with <code>equals</code>, unequal (i.e.,
+ * different) elements that are equal according to the order can exist in a set. In this case,
+ * elements that are added first will have a higher rank than elements of the same order that are
+ * added later.
+ *
+ * @param <T> the type of elements maintained by this set
+ */
+public interface OrderStatisticSet<T> extends SortedSet<T> {
```

Comment:
Does it make sense for all `OrderStatisticSet`s to extend `NavigableSet`? Or does it at least make sense for `SkipList`?

Reply:
It definitely makes sense for the `SkipList`.
Order statistic tree implementations are also able to get the floor/ceil values easily,
but I'm not sure whether this property can be derived from the interface, in general.

Do  you have an opinion on that?

Reply:
I know less about these data structures than you, so it is up to you to decide whether this makes sense.

Reply:
OK, so how about that notion:

Every collection that is able to get the rank of an element in an ordered set has to determine the position of that element in the list.
If it can do that, it can also determine where an element in an ordered set *should* be, and thus determine its neighbors.

Thus, every implementation of `OrderStatisticSet` should also be able to implement `NavigableSet` efficiently.